panther_battery/test/test_adc_battery.cpp

```diff
@@ -0,0 +1,247 @@
+
+  // variable values
+  EXPECT_FLOAT_EQ(expected_voltage, battery_state_.voltage);
+  EXPECT_FLOAT_EQ(expected_current, battery_state_.current);
+  EXPECT_FLOAT_EQ(expected_percentage, battery_state_.percentage);
+  EXPECT_FLOAT_EQ(expected_percentage * 20.0f, battery_state_.charge);
+
+  EXPECT_TRUE(battery_state_.present);
+  EXPECT_EQ(power_supply_status, battery_state_.power_supply_status);
+  EXPECT_EQ(power_supply_health, battery_state_.power_supply_health);
+}
+
+TEST_F(TestADCBattery, BatteryPresent)
+{
+  battery_temp_raw_ = 1.0;
+  EXPECT_TRUE(battery_->Present());
+  battery_temp_raw_ = 3.03;
+  EXPECT_FALSE(battery_->Present());
+  battery_temp_raw_ = 3.5;
+  EXPECT_FALSE(battery_->Present());
+}
+
+TEST_F(TestADCBattery, BatteryMsgUnknown)
+{
+  auto stamp = rclcpp::Time(0);
+  battery_->Reset(stamp);
+  battery_state_ = battery_->GetBatteryMsg();
+
+  TestDefaultBatteryStateMsg(
+    BatteryStateMsg::POWER_SUPPLY_STATUS_UNKNOWN, BatteryStateMsg::POWER_SUPPLY_HEALTH_UNKNOWN);
+}
+
+TEST_F(TestADCBattery, BatteryMsgValues)
+{
+  const auto voltage_factor = 25.04255;
+  const auto current_factor = 20.0;
+  const auto charge_factor = 2.5;
+  const auto V_bat_full = 41.4;
+  const auto V_bat_min = 32.0;
+
+  UpdateBattery(1.5, 0.01, 1.5, 0.5, false);
+  float expected_voltage = 1.5 * voltage_factor;
+  float expected_percentage = (expected_voltage - V_bat_min) / (V_bat_full - V_bat_min);
+  float expected_temp = 28.875206;
+  float expected_current = -(0.01 * current_factor) + (0.5 * charge_factor);
+  TestBatteryStateMsg(
+    expected_voltage, expected_current, expected_temp, expected_percentage,
+    BatteryStateMsg::POWER_SUPPLY_STATUS_DISCHARGING, BatteryStateMsg::POWER_SUPPLY_HEALTH_GOOD);
+
+  // check if values are smoothed out with moving average.
+  UpdateBattery(1.6, 0.02, 1.4, 0.4, false);
+  const float voltage_mean = (1.5 + 1.6) / 2.0;
+  const float current_mean = (0.01 + 0.02) / 2.0;
+  const float charge_mean = (0.5 + 0.4) / 2.0;
+  expected_voltage = voltage_mean * voltage_factor;
+  expected_percentage = (expected_voltage - V_bat_min) / (V_bat_full - V_bat_min);
+  expected_temp = 30.306725;
+  expected_current = -(current_mean * current_factor) + (charge_mean * charge_factor);
+  TestBatteryStateMsg(
+    expected_voltage, expected_current, expected_temp, expected_percentage,
+    BatteryStateMsg::POWER_SUPPLY_STATUS_DISCHARGING, BatteryStateMsg::POWER_SUPPLY_HEALTH_GOOD);
+
+  // Raw battery message should depend only on last readings
+  battery_state_ = battery_->GetBatteryMsgRaw();
+  expected_voltage = 1.6 * voltage_factor;
+  expected_percentage = (expected_voltage - V_bat_min) / (V_bat_full - V_bat_min);
+  expected_temp = 31.738245;
+  expected_current = -(0.02 * current_factor) + (0.4 * charge_factor);
+  TestBatteryStateMsg(
+    expected_voltage, expected_current, expected_temp, expected_percentage,
+    BatteryStateMsg::POWER_SUPPLY_STATUS_DISCHARGING, BatteryStateMsg::POWER_SUPPLY_HEALTH_GOOD);
```

Comment:
It would be better to also move raw values passed to the UpdateBattery to parameters